Firestore/core/src/core/event_listener.h

```diff
@@ -86,7 +83,19 @@ class AsyncEventListener
   void Mute();
 
  private:
-  std::atomic<bool> muted_;
+  // PORTING NOTE: Android uses a volatile here but that's not enough in C++.
+  //
+  // In C++, the user can call `ListenerRegistration::Remove` (which calls
+  // `Mute`) and then immediately delete the state backing the listener. Using
+  // a mutex here instead of an atomic ensures that `Mute` won't return until
+  // it's safe to delete the state backing a listener. In Java this is safe
+  // because the state backing the listener is garbage collected so it doesn't
+  // matter if the mute is concurrent with a callback.
+  //
+  // Use a recursive mutex instead of `std::mutex` to avoid deadlock in the case
```

Comment:
Is there a test case dedicated to verifying no deadlock if Remove() is called by a callback? I didn't see one, but I may have missed it. I realize that you exercise this code path in the new unit test you added but it should have a dedicated unit test being that it would be a "guarantee" that would be easy to break unintentionally.

Reply:
I've added a test.